python/cuml/common/doc_utils.py

```diff
@@ -0,0 +1,423 @@
+            elif par in ['X', 'y'] and par not in skip_parameters:
+                docstring_wrapper.__doc__ += \
+                    _parameters_docstrings[X].format(name=par,
+                                                     shape=X_shape)
+
+            # convert_dtype requires some magic to distinguish
+            # whether we use the fit version or the version
+            # for the other methods.
+            elif par == 'convert_dtype' and par not in skip_parameters:
+                if not convert_dtype_cast:
+                    if func.__name__ == 'fit':
+                        k = 'convert_dtype_fit'
+                    else:
+                        k = 'convert_dtype_other'
+
+                    docstring_wrapper.__doc__ += \
+                        _parameters_docstrings[k].format(
+                            default=params['convert_dtype'].default,
+                            func_name=func.__name__
+                        )
+
+                else:
+                    docstring_wrapper.__doc__ += \
+                        _parameters_docstrings['convert_dtype_single'].format(
+                            default=params['convert_dtype'].default,
+                            dtype=convert_dtype_cast
+                        )
+
+            # All other parameters only take a default (for now).
+            else:
+                if par in _simple_params:
+                    docstring_wrapper.__doc__ += \
+                        _parameters_docstrings[par].format(
+                            default=params[par].default
+                        )
+            docstring_wrapper.__doc__ += '\n\n'
+
+        # Add return section header if needed, no option to skip currently.
+        if(return_values):
+            docstring_wrapper.__doc__ += \
+                '\nReturns \n ---------- \n'
+
+            # convenience call to allow users to pass a single return
+            # value as a dictionary instead of a list of dictionaries
+            rets = [return_values] if not isinstance(return_values, list) \
+                else return_values
+
+            # process each entry in the return_values
+            # auto naming of predicted variable names will be a
+            # future improvement
+            for ret in rets:
+                if ret['type'] in _return_values_docstrings:
+                    key = ret['type']
+                    # non custom types don't take the type parameter
+                    del ret['type']
+                else:
+                    key = 'custom_type'
+
+                # ret is already a dictionary, we just use it for the named
+                # parameters
+                docstring_wrapper.__doc__ += \
+                    _return_values_docstrings[key].format(
+                        **ret
+                    )
+                docstring_wrapper.__doc__ += '\n\n'
+
+        return docstring_wrapper
+    return deco
+
+
+def insert_into_docstring(parameters=False,
```

Comment:
It would be better to use named placed holders instead of indexed ones. For example, you could replace
```
        Parameters
        ----------
        X : {}
        n_neighbors : Integer
        ...
```
with 
```
        Parameters
        ----------
        {X}
        n_neighbors : Integer
        ...
```
And then you would use the decorator with a dictionary instead of a list. This:
```python
@insert_into_docstring(
    parameters=[('dense', '(n_samples, n_features)')],
    return_values=[
        ('dense', '(n_samples, n_features)'),
        ('dense', '(n_samples, n_features)')
    ])
```
becomes:
```python
@insert_into_docstring(
    parameters={ "X": ('dense', '(n_samples, n_features)') },
    return_values={
        "distances": ('dense', '(n_samples, n_features)'),
        "indices": ('dense', '(n_samples, n_features)')
    })
```

This has a few benefits:

1. Order is not important. If a parameter is added in the future, the doc string will still be accurate
2. The generator would now be responsible for writing the entire line. This makes it easier to match the indent
3. Overall, I think it makes the decorator easier to understand.

Reply:
That is the version I had implemented originally, the main issue being that things start becoming very verbose very quickly, while I'm not sure if we need that much flexibility

Reply:
Fair enough, I definitely prefer using a map, but I dont think its a deal breaker. If you feel strongly about using arrays over maps, we can still implement the correct indentation. It would just take a bit more work.